Refactor: introduce WC_Payments_API_Setup_Intention for setup intents

htdat opened the pull request:
Fixes #2422

#### Changes proposed in this Pull Request

- Create a new abstraction class `WC_Payments_API_Abstract_Intention`, and `WC_Payments_API_Setup_Intention` for setup intents to extend on this abstract class.
- Update the old payment intent class to `WC_Payments_API_Payment_Intention`, that now extends the abstract class. Move all shared getters and properties from this payment intent class to the abstract class.
- Fix test failures.
- For `WC_Payments_API_Abstract_Intention`, I also introduced the `get_payment_method_type` to assist the process of extracting the payment method type (`card`, `sofort`, `card_present`, etc) easily.

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

Basically, this refactor will touch to multiple flows, and ideally, reviewers/testers would test [all critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows) regarding the payment process in the checkout or pay-for-order pages. That would include: 

- Flows: traditional (card), UPE, split-UPE, deferred UPE.
- Subscription products: free-trial (zero-amount orders), and normal (total > 0) orders.
- Checkout pages: classic and block.

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

*

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] NON-RELEVANT Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [X] NO NEED - USING THE CURRENT CRITICAL FLOWS IS ENOUGH - Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [X] NO NEED - USING THE CURRENT CRITICAL FLOWS IS ENOUGH _ Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [X] NO NEED Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.
- [ ] Monitor GlobalStep testings, especially check PHP error log wp-content/debug.log from their test sites.

htdat committed:
Initial refactor

- Create  WC_Payments_API_Abstract_Intention
- Rename WC_Payments_API_Payment_Intention and extends the abstract.

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `6586` or branch name `refactor/2422-object-support-setup-intents` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=refactor/2422-object-support-setup-intents) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 473037d9ac172f3a1776d02bc427ed48c7c124d7
- Build time: 2023-07-21 15:39:09 UTC

_Note: the build is updated when a new commit is pushed to this PR._

htdat committed:
Clean up the payment intent and abstract intent object

- Leave payment intent object with the properties just exists in it.

htdat committed:
Add WC_Payments_API_Setup_Intention class

htdat committed:
Add function deserialize_setup_intention_object_from_array

htdat committed:
Update gateway classes to use new setup intent object.

htdat committed:
Add Setup_Intent_Status and fix get_status for setup_intent

htdat committed:
Fix tests

htdat committed:
Fix tests for webhook processing

htdat committed:
Merge branch 'develop' into refactor/2422-object-support-setup-intents

htdat committed:
Use WC_Helper_Intention::create_intention for test_create_terminal_intent_success

htdat committed:
Remove the unnecessary part for preparing intent in the webhook processing

Left-over from d6036621955f7b6d3a0068c221ebfe6f40a28bc1

htdat committed:
Optimize WC_Payments_Order_Service::get_intent_data

htdat committed:
Add changelog

htdat committed:
Change to deserialize_payment_intention_object_from_array to specify for payment intents

htdat committed:
Update  deserialize_payment_intention_object_from_array in tests

Due to the change in e0d69dfd47be06db3f3113b204695131ebf6b1f8

htdat committed:
Fix metadata can be empty object

- Reason https://github.com/Automattic/woocommerce-payments/pull/419/commits/c2c8438c3ed7be6d604435e059209fb87fb6d0c4.

- Metadata in setup intents in UPE is empty now  https://github.com/Automattic/woocommerce-payments/issues/6575

htdat committed:
Fix metadata can be empty object

- Reason https://github.com/Automattic/woocommerce-payments/pull/419/commits/c2c8438c3ed7be6d604435e059209fb87fb6d0c4.

- Metadata in setup intents in UPE is empty now  https://github.com/Automattic/woocommerce-payments/issues/6575

htdat committed:
Avoid the issue due to missing setup intent IDs attached to the order.

Then to make it work properly in deserialize_setup_intention_object_from_array

 Issue https://github.com/Automattic/woocommerce-payments/issues/6576

htdat committed:
Update to use payment_method_options

htdat committed:
Update create_setup_intent_ajax to use the setup intent object

htdat committed:
Merge branch 'develop' into refactor/2422-object-support-setup-intents

htdat committed:
Add comments for intent classes.

htdat commented:
Ready to review!

RadoslavGeorgiev commented on includes/constants/class-setup-intent-status.php:
```diff
@@ -0,0 +1,27 @@
+ *
+ * @psalm-immutable
+ */
+class Setup_Intent_Status extends Base_Constant {
```
AFAICT the only _payment_-intent-specific status is `requires_capture`, which has very specific handling. With that in mind, would you consider using a single `Intent_Status` ENUM throughout the gateway? This way, most code would be agnostic to the type of intent being used. 

Also, right now, the gateway uses `WC_Payment_Gateway_WCPay::SUCCESSFUL_INTENT_STATUS` regardless.

htdat replied:
Good point. Did that here 1f10437ea85e15ee4cabcbb590a7ca6b635f8819 and created this issue to refactor to `Intent_Status` later https://github.com/Automattic/woocommerce-payments/issues/6773

RadoslavGeorgiev commented on bin/run-tests.sh:
```diff
@@ -13,10 +13,10 @@ while getopts ':w' OPTION; do
+#echo "Installing the test environment..."
+#
+#docker-compose exec -u www-data wordpress \
+#	/var/www/html/wp-content/plugins/woocommerce-payments/bin/install-wp-tests.sh
```
Is this intentionally left commented?

htdat replied:
Good catch! Revert 3db3b12ce504a25e7efa4d8a391c15185fa2e2e6

RadoslavGeorgiev commented on includes/core/server/request/class-create-and-confirm-setup-intention.php:
```diff
@@ -83,4 +84,14 @@ public function set_payment_method( string $payment_method_id ) {
+	 * @return mixed           Either the same response, or the correct object.
+	 */
+	public function format_response( $response ) {
+		return WC_Payments::get_payments_api_client()->deserialize_setup_intention_object_from_array( $response );
```
I'm not sure how this managed to slip through the cracks previously, but the dependency on `WC_Payments::get_payments_api_client()` is not needed. The `Request` class already contains the API client, so a simple call to the dependency should be better.

Could you please change that in all requests? If not within this PR, let's open an issue for next sprint to fix that.

htdat replied:
Good point! Updated b0130ee2cdf0e2a6aeddd587f4d5007ed58d5e94

RadoslavGeorgiev commented on includes/class-wc-payments-order-service.php:
```diff
@@ -1438,32 +1438,25 @@ private function update_order_status( $order, $order_status, $intent_id = '' ) {
+			'payment_method_type' => $intent->get_payment_method_type(),
+		];
+
+		if ( is_a( $intent, WC_Payments_API_Payment_Intention::class ) ) {
```
```suggestion
		if ( $intent instanceof WC_Payments_API_Payment_Intention ) ) {
```

Let's use `instanceof` instead of `is_a()`. It's functionally the same, but VS Code understands it a bit better. 
<img width="936" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/5311119/c653c684-29cb-45fd-99fb-b2849a8b5f75">
<img width="849" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/5311119/af753c45-f344-436b-a4ae-e16b03d8556a">

htdat replied:
PHPStorm seems smarter :D 

![Markup on 2023-07-14 at 14:21:39](https://github.com/Automattic/woocommerce-payments/assets/10045087/6b4ca50a-37aa-4c4b-bd7b-96452b407a1a)

Anyway, I've made the change: c9c09c1461876f51f81a4a54925b11b9455f47a2

RadoslavGeorgiev commented on includes/wc-payment-api/class-wc-payments-api-client.php:
```diff
@@ -2121,7 +2122,7 @@ private function add_order_info_to_charge_object( $charge_id, $object ) {
 		// Add order information to the `$transaction`.
 		// If the order couldn't be retrieved, return an empty order.
-		$object['order'] = null;
+		$object['order'] = [];
```
Could you provide some reasoning here? Why is an empty order needed instead of just `null`? I [see the linked issue](https://github.com/Automattic/woocommerce-payments/pull/6586/commits/941bf3970c66c8080f2621080f3df2e28c002b5a), but I don't understand how having an array is better.

htdat replied:
@RadoslavGeorgiev  - Good point, I would be more verbose for this simple change. 

First, if testing this PR with replicate steps in #6576, here is the error:

<details>

<summary>Click to see the full error with Stack trace</summary>

```
[14-Jul-2023 08:25:03 UTC] PHP Fatal error:  Uncaught TypeError: Argument 12 passed to WC_Payments_API_Setup_Intention::__construct() must be of the type array, null given, called in /var/www/html/wp-content/plugins/woocommerce-payments/includes/wc-payment-api/class-wc-payments-api-client.php on line 2313 and defined in /var/www/html/wp-content/plugins/woocommerce-payments/includes/wc-payment-api/models/class-wc-payments-api-setup-intention.php:38
Stack trace:
#0 /var/www/html/wp-content/plugins/woocommerce-payments/includes/wc-payment-api/class-wc-payments-api-client.php(2313): WC_Payments_API_Setup_Intention->__construct('seti_1NThJDR67t...', 'cus_NczxswVwvaT...', NULL, Object(DateTime), 'requires_paymen...', 'seti_1NThJDR67t...', Array, Array, Array, Array, Array, NULL)
#1 /var/www/html/wp-content/plugins/woocommerce-payments/includes/core/server/request/class-create-setup-intention.php(76): WC_Payments_API_Client->deserialize_setup_intention_object_from_array(Array)
#2 /var/www/html/wp-content/plugins/woocommerce-payments/includes/core/server/class- in /var/www/html/wp-content/plugins/woocommerce-payments/includes/wc-payment-api/models/class-wc-payments-api-setup-intention.php on line 38
```

</details>

It's a chain of actions to see an error like that. I also it's not an intention to give a `null` here so changing it to an array makes a lot of sense with the comment too: 

1, Method `deserialize_payment_intention_object_from_array` calls this https://github.com/Automattic/woocommerce-payments/blob/e3f1820ac05b9932a94ce1900d070363078c7eff/includes/wc-payment-api/class-wc-payments-api-client.php#L2310

2, Due to the bug #6576, `$order` here is returned `false`, i.e. no order with the provided setup intent ID. https://github.com/Automattic/woocommerce-payments/blob/e3f1820ac05b9932a94ce1900d070363078c7eff/includes/wc-payment-api/class-wc-payments-api-client.php#L2092-L2096

3, In the next iteration, `$object['order']` will be `null` in `add_order_info_to_object`. 

Then, finally the `$order` is used in step 1 above is `null`, not an array as expected.

RadoslavGeorgiev reviewed (dismissed):
The changes look great, @htdat! 👏 

I've got a couple of minor comments, but the rest of the code looks excellent. I haven't performed thorough manual testing yet, as there are quite a few scenarios to test, but nothing looks dangerous right now.

htdat committed:
Merge branch 'develop' into refactor/2422-object-support-setup-intents

htdat committed:
Revert accidental changes for run-tests.sh

htdat committed:
Remove Setup_Intent_Status and come back to use Payment_Intent_Status

htdat committed:
Change to use the dependency for WC_Payments_API_Client in request classes.

htdat committed:
Feedback in using instanceof rather than is_a

htdat committed:
Add param type for get_intent_data

htdat committed:
Merge branch 'develop' into refactor/2422-object-support-setup-intents

htdat commented:
@Automattic/gamma - ready to be reviewed again after addressing feedback from @RadoslavGeorgiev.

At the moment, `E2E test failure regarding dispute` and `PHP test failure regarding woopay request` are known issues p1689305403447819-slack-CGGCLBN58

zmaglica committed:
Merge branch 'develop' into refactor/2422-object-support-setup-intents

zmaglica commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -1182,7 +1183,8 @@ public function process_payment_for_order( $cart, $payment_information, $schedul
 
-				$intent_meta_order_id_raw = ! empty( $intent['metadata'] ) ? $intent['metadata']['order_id'] ?? '' : '';
+				$intent_metadata          = $intent->get_metadata();
+				$intent_meta_order_id_raw = ! empty( $intent_metadata ) ? $intent_metadata['order_id'] ?? '' : '';
```
Nitpick. You could simplify this line by doing only $intent_metadata['order_id'] ?? ''. You don't have to do an empty check since the null coalescing operator will do that for you and you are already doing it.

htdat replied:
Addressed fe80244

zmaglica commented on includes/wc-payment-api/class-wc-payments-api-client.php:
```diff
@@ -2229,31 +2230,37 @@ private function deserialize_charge_object_from_array( array $charge_array ) {
+			? $raw_metadata
+			: [];
+		$charge_array           = 0 < $intention_array['charges']['total_count'] ? end( $intention_array['charges']['data'] ) : null;
+		$next_action            = ! empty( $intention_array['next_action'] ) ? $intention_array['next_action'] : [];
```
Can you use `$intention_array['next_action'] ?? []` here? It is a small nitpick but makes the code a bit easier to read.

htdat replied:
Addressed fe80244

zmaglica commented on includes/wc-payment-api/class-wc-payments-api-client.php:
```diff
@@ -2268,6 +2275,52 @@ public function deserialize_intention_object_from_array( array $intention_array
+		$metadata               = is_array( $raw_metadata ) && ! empty( $raw_metadata )
+			? $raw_metadata
+			: [];
+		$next_action            = ! empty( $intention_array['next_action'] ) ? $intention_array['next_action'] : [];
```
Same comment as above regarding the null coalescing operator for this type of checks

htdat replied:
Addressed fe80244

zmaglica reviewed (changes requested):
Left a couple of nitpicks. The overall code looks good.

htdat committed:
Merge branch 'develop' into refactor/2422-object-support-setup-intents

htdat committed:
Address feedback, use null coalescing

htdat committed:
TRY - NO MERGE

htdat committed:
Add Get_Setup_Intention request class

htdat committed:
Replace Request::get with Get_Setup_Intention

- Also fixing previous merge from develop

htdat committed:
Fixing tests

Due to:

- Adding Get_Setup_Intention class
- Also fixing previous merge from develop

htdat committed:
Merge branch 'develop' into refactor/2422-object-support-setup-intents

htdat commented:
👋 @zmaglica Ready to review again. Major changes from your last review: 

- fe8024495 Address your feedback, use null coalescing
- f98a63183 This merge includes your PR 6f71c55cb, which also touches `get_setup_intent`. Therefore, I needed to implement the following changes. 
- 1103024c6 Add Get_Setup_Intention request class, and take advantage of the new model class `WC_Payments_API_Setup_Intention`. 
- 05956b9f3 Replace Request::get for get_setup_intent with Get_Setup_Intention
- 161608654 Fixing tests

zmaglica commented on includes/core/server/request/class-get-setup-intention.php:
```diff
@@ -0,0 +1,67 @@
+	 *
+	 * @throws Invalid_Request_Parameter_Exception
+	 */
+	public function __construct( WC_Payments_API_Client $api_client, WC_Payments_Http_Interface $http_interface, string $setup_intent_id ) {
```
We already support id argument via the default create method:

```php
public static function create( $id = null ) {
		return WC_Payments::create_request( static::class, $id );
}
```

and default constructor looks like this:

```php
	public function __construct( WC_Payments_API_Client $api_client, WC_Payments_Http_Interface $http_interface, $id = null ) {
		$this->api_client     = $api_client;
		$this->http_interface = $http_interface;

		if ( method_exists( $this, 'set_id' ) ) {
			if ( null !== $id ) {
				$this->set_id( $id );
			} else {
				throw new Invalid_Request_Parameter_Exception( 'This request requires an item ID.', 'wcpay_core_invalid_request_parameter_missing_id' );
			}
		}
	}
```

In other words, your request class only needs to define `set_id` function where you can validate id by calling `validate_stripe_id`. You don't have to override the default constructor.

zmaglica commented on includes/core/server/request/class-get-setup-intention.php:
```diff
@@ -0,0 +1,67 @@
+	 * Formats the response from the server.
+	 *
+	 * @param  mixed $response The response from `WC_Payments_API_Client::request`.
+	 * @return mixed           Either the same response, or the correct object.
```
You can say that the return type will be `WC_Payments_API_Setup_Intention` and get all typrhints and other goodies.

htdat committed:
Address review: add type hint, and use supported magic id

htdat commented:
@zmaglica - Thanks for the last feedback. Addressed in 54835b1a6c7ece6ed05dfa3036e324923b5d48b5

Ready to review again.

zmaglica reviewed (approved):
LGTM! Good job on converting some of the requests to the standalone request classes.

zmaglica committed:
Merge branch 'develop' into refactor/2422-object-support-setup-intents

htdat committed:
Merge branch 'develop' into refactor/2422-object-support-setup-intents
